src/core/datastructures/bitset.cpp

```diff
@@ -0,0 +1,268 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/core/datastructures/bitset.h>
+
+#include <warn/push>
+#include <warn/ignore/all>
+#include <roaring.hh>
+#include <warn/pop>
+
+namespace inviwo {
+
+BitSet::BitSetIterator::~BitSetIterator() = default;
+
+BitSet::BitSetIterator::BitSetIterator(const BitSet& b, bool exhausted)
+    : it_(std::make_unique<RoaringIt>(*b.roaring_, exhausted)) {}
+
+BitSet::BitSetIterator::BitSetIterator(const RoaringIt& it)
+    : it_(std::make_unique<RoaringIt>(it)) {}
+
+BitSet::BitSetIterator::BitSetIterator(const BitSetIterator& rhs)
+    : it_(std::make_unique<RoaringIt>(*rhs.it_)) {}
+
+auto BitSet::BitSetIterator::operator=(const BitSetIterator& rhs) -> BitSetIterator& {
+    if (this != &rhs) {
+        *it_ = *rhs.it_;
+    }
+    return *this;
+}
+
+BitSet::BitSetIterator& BitSet::BitSetIterator::operator++() {
+    it_->operator++();
+    return *this;
+}
+
+BitSet::BitSetIterator BitSet::BitSetIterator::operator++(int) {
+    BitSet::BitSetIterator it(*it_);
+    it_->operator++();
+    return it;  // <- attempting to reference deleted function BitSetIterator(const BitSetIterator&)
+}
+
+BitSet::BitSetIterator& BitSet::BitSetIterator::operator--() {
+    it_->operator--();
+    return *this;
+}
+
+BitSet::BitSetIterator BitSet::BitSetIterator::operator--(int) {
+    BitSetIterator it(*it_);
+    it_->operator--();
+    return it;
+}
+
+auto BitSet::BitSetIterator::operator*() const -> value_type { return **it_; }
+
+bool BitSet::BitSetIterator::operator==(const BitSetIterator& rhs) const {
+    return it_->operator==(*rhs.it_);
+}
+
+bool BitSet::BitSetIterator::operator!=(const BitSetIterator& rhs) const {
+    return it_->operator!=(*rhs.it_);
+}
+
+BitSet::BitSet()
```

Comment:
could also add a private conversion constructor from roaring::Roaring, would make some of the functions easier to implement I think, would avoid having to create some empty bitsets

Reply:
However, internally Roaring always creates an empty bitmap beforehand. Compare
```c++
    Roaring() : roaring{} {
        // The empty constructor roaring{} silences warnings from pedantic static analyzers.
        api::roaring_bitmap_init_cleared(&roaring);
    }

    /**
     * Construct a bitmap from a list of integer values.
     */
    Roaring(size_t n, const uint32_t *data) : Roaring() {
        api::roaring_bitmap_add_many(&roaring, n, data);
    }
```

Reply:
I was mostly thinking it might simplify our side a bit.